File: crimson-zig/src/typo/runtime.zig
const std = @import("std");
const rng_callers = @import("../rng_caller_static.zig");
const creatures_mod = @import("../runtime/creatures.zig");
const player_runtime = @import("../runtime/player.zig");
const state_mod = @import("../runtime/state.zig");
const spawn_mod = @import("../runtime/spawn.zig");
const typo_names = @import("names.zig");
const typo_player = @import("player.zig");
const typo_spawns = @import("spawns.zig");

fn typeclickSfx(
    state: *state_mod.GameplayState,
    caller: rng_callers.Caller,
) state_mod.SfxId {
    if ((state.rng.randTagged(caller) & 1) == 0) return .ui_typeclick_01;
    return .ui_typeclick_02;
}

pub fn applyCharCommand(state: *state_mod.GameplayState, ch: u8) void {
    if (ch == 0) return;
    state.typo.typing.pushChar(ch);
    state.sfx_queue.append(typeclickSfx(state, rng_callers.typo_gameplay_typeclick_char));
}

pub fn applyBackspaceCommand(state: *state_mod.GameplayState) void {
    state.typo.typing.backspace();
    state.sfx_queue.append(typeclickSfx(state, rng_callers.typo_gameplay_typeclick_backspace));
}

pub fn applySubmitCommand(
    state: *state_mod.GameplayState,
    creatures: *const creatures_mod.CreaturePool,
) void {
    const entered = state.typo.typing.submit(false) orelse return;
    state.sfx_queue.append(.ui_typeenter);

    var active_mask = [_]bool{false} ** creatures_mod.max_creatures;
    for (creatures.entries, 0..) |entry, idx| {
        active_mask[idx] = entry.active;
    }

    if (std.mem.eql(u8, entered, "reload")) {
        state.typo.pending_reload = true;
        return;
    }

    if (state.typo.names.findByName(entered, active_mask[0..])) |idx| {
        const creature = creatures.entries[idx];
        if (creature.active) {
            state.typo.pending_fire_target_active = true;
            state.typo.pending_fire_target_x = creature.pos.x;
            state.typo.pending_fire_target_y = creature.pos.y;
            state.typo.typing.match_count += 1;
        }
    }
}

pub fn beforeStep(
    state: *state_mod.GameplayState,
    players: []state_mod.PlayerState,
) void {
    for (players) |*player| {
        typo_player.enforceTypoPlayerFrame(player, state);
    }
}

pub fn transformPrimaryInput(
    state: *state_mod.GameplayState,
    input: player_runtime.GameInput,
) player_runtime.GameInput {
    var transformed = input;
    transformed.move_x = 0.0;
    transformed.move_y = 0.0;
    transformed.flags.fire_down = false;
    transformed.flags.fire_pressed = false;
    transformed.flags.reload_pressed = false;
    transformed.flags.reload_down = false;

    if (state.typo.pending_fire_target_active) {
        transformed.aim_x = state.typo.pending_fire_target_x;
        transformed.aim_y = state.typo.pending_fire_target_y;
        transformed.flags.fire_down = true;
        transformed.flags.fire_pressed = true;
    }
    if (state.typo.pending_reload) {
        transformed.flags.reload_pressed = true;
    }

    state.typo.pending_fire_target_active = false;
    state.typo.pending_reload = false;
    return transformed;
}

pub fn midStep(
    state: *state_mod.GameplayState,
    players: []const state_mod.PlayerState,
    creatures: *creatures_mod.CreaturePool,
    elapsed_before_ms: f32,
    dt_sim_ms: f32,
    world_size: f32,
) void {
    const batch = typo_spawns.tickTypoSpawns(
        @intFromFloat(elapsed_before_ms),
        state.typo.spawn_cooldown_ms,
        @intFromFloat(dt_sim_ms),
        @intCast(players.len),
        world_size,
        world_size,
    );
    state.typo.spawn_cooldown_ms = batch.cooldown_ms;

    var active_mask = [_]bool{false} ** creatures_mod.max_creatures;
    for (creatures.entries, 0..) |entry, idx| {
        active_mask[idx] = entry.active;
    }

    const score_xp: i32 = if (players.len > 0) players[0].experience else 0;
    var dict_storage: [typo_names.max_dictionary_words][]const u8 = undefined;
    for (0..state.typo.dictionary_word_count) |idx| {
        dict_storage[idx] = state.typo.dictionaryWordSlice(idx);
    }
    var hs_storage: [typo_names.max_highscore_names][]const u8 = undefined;
    for (0..state.typo.highscore_name_count) |idx| {
        hs_storage[idx] = state.typo.highscoreNameSlice(idx);
    }

    for (batch.slice()) |call| {
        // creature_spawn_tinted allocates via creature_alloc_slot, which seeds
        // phase_seed = float(crt_rand() & 0x17f) before the heading/size draws.
        const phase_seed = @as(f32, @floatFromInt(state.rng.randTagged(rng_callers.creature_alloc_slot_phase_seed) & 0x17f));
        const heading = @as(f32, @floatFromInt(state.rng.randTagged(rng_callers.creature_spawn_tinted_heading) % 314)) * 0.01;
        var size = @as(f32, @floatFromInt(state.rng.randTagged(rng_callers.creature_spawn_tinted_size) % 20 + 47));
        var flags: i32 = 0;
        var move_speed: f32 = 1.7;
        if (call.type_id == .spider_sp1 or call.type_id == .spider_sp2) {
            flags |= spawn_mod.CreatureFlags.ai7_link_timer;
            move_speed *= 1.2;
            size *= 0.8;
        }

        const creature_idx = creatures.spawnInit(.{
            .origin_template_id = 0,
            .pos = .{ .x = call.pos_x, .y = call.pos_y },
            .heading = heading,
            .phase_seed = phase_seed,
            .type_id = call.type_id,
            .flags = @bitCast(flags),
            .ai_mode = .chase_player,
            .health = 1.0,
            .max_health = 1.0,
            .move_speed = move_speed,
            .reward_value = 1.0,
            .size = size,
            .contact_damage = 100.0,
        });
        active_mask[creature_idx] = true;
        _ = state.typo.names.assignRandom(
            creature_idx,
            &state.rng,
            score_xp,
            active_mask[0..],
            dict_storage[0..state.typo.dictionary_word_count],
            hs_storage[0..state.typo.highscore_name_count],
        );
    }
}

pub fn postStep(state: *state_mod.GameplayState) void {
    state.bonuses.weapon_power_up = 0.0;
    state.bonuses.reflex_boost = 0.0;
    state.time_scale_active = false;
}

File: src/crimson/typo/runtime.py
from __future__ import annotations

from typing import TYPE_CHECKING

import msgspec

from grim.geom import Vec2
from grim.sfx_map import SfxId

from ..creatures.spawn import CreatureAiMode, CreatureFlags, CreatureInit, CreatureTypeId
from ..rng_caller_static import RngCallerStatic
from ..sim.input import PlayerInput
from ..sim.input_providers import TypoBackspaceCommand, TypoCharCommand, TypoSubmitCommand
from ..sim.world_state import WorldState
from .player import enforce_typo_player_frame
from .spawns import tick_typo_spawns

if TYPE_CHECKING:
    from ..sim.sessions import MidStepContext, PostStepContext


def _require_single_player_typo(command) -> None:
    if int(command.player_index) != 0:
        raise RuntimeError("Typ-o Shooter commands are single-player only")


def _typeclick_sfx(world: WorldState, *, caller: RngCallerStatic) -> SfxId:
    if (world.state.rng.rand_tagged(caller) & 1) == 0:
        return SfxId.UI_TYPECLICK_01
    return SfxId.UI_TYPECLICK_02


def apply_typo_command(world: WorldState, command: TypoCharCommand | TypoBackspaceCommand | TypoSubmitCommand) -> None:
    _require_single_player_typo(command)
    typo = world.state.typo
    typing = typo.typing

    match command:
        case TypoCharCommand(ch=ch):
            if ch:
                typing.push_char(str(ch))
                world.state.sfx_queue.append(
                    _typeclick_sfx(world, caller=RngCallerStatic.TYPO_GAMEPLAY_TYPECLICK_CHAR),
                )
        case TypoBackspaceCommand():
            typing.backspace()
            world.state.sfx_queue.append(
                _typeclick_sfx(world, caller=RngCallerStatic.TYPO_GAMEPLAY_TYPECLICK_BACKSPACE),
            )
        case TypoSubmitCommand():
            if not typing.text:
                return
            world.state.sfx_queue.append(SfxId.UI_TYPEENTER)
            active_mask = [bool(entry.active) for entry in world.creatures.entries]
            target_idx = typo.names.find_by_name(typing.text, active_mask=active_mask)
            entered = typing.submit(matched=target_idx is not None)
            typo.pending_fire_target = None
            typo.pending_reload = False
            if entered is None:
                return
            if target_idx is not None:
                creature = world.creatures.entries[int(target_idx)]
                if creature.active:
                    typo.pending_fire_target = Vec2(float(creature.pos.x), float(creature.pos.y))
                return
            if entered == "reload":
                typo.pending_reload = True
        case _:
            raise RuntimeError(f"unhandled Typ-o command: {type(command).__name__}")


def typo_before_step(world: WorldState) -> None:
    for player in world.players:
        enforce_typo_player_frame(player, state=world.state)


def typo_mid_step(ctx: MidStepContext) -> None:
    typo = ctx.world.state.typo
    cooldown, spawns = tick_typo_spawns(
        elapsed_ms=int(ctx.elapsed_before_ms),
        spawn_cooldown_ms=int(typo.spawn_cooldown_ms),
        frame_dt_ms=int(ctx.dt_sim_ms),
        player_count=len(ctx.world.players),
        world_width=float(ctx.world_size),
        world_height=float(ctx.world_size),
    )
    typo.spawn_cooldown_ms = int(cooldown)
    for call in spawns:
        # creature_spawn_tinted allocates via creature_alloc_slot, which seeds
        # phase_seed = float(crt_rand() & 0x17f) before the heading/size draws.
        phase_seed = float(ctx.world.state.rng.rand_tagged(RngCallerStatic.CREATURE_ALLOC_SLOT_PHASE_SEED) & 0x17F)
        heading = float(ctx.world.state.rng.rand_tagged(RngCallerStatic.CREATURE_SPAWN_TINTED_HEADING) % 314) * 0.01
        size = float(ctx.world.state.rng.rand_tagged(RngCallerStatic.CREATURE_SPAWN_TINTED_SIZE) % 20 + 47)
        flags = CreatureFlags(0)
        move_speed = 1.7
        if int(call.type_id) in (int(CreatureTypeId.SPIDER_SP1), int(CreatureTypeId.SPIDER_SP2)):
            flags |= CreatureFlags.AI7_LINK_TIMER
            move_speed *= 1.2
            size *= 0.8

        creature_idx = ctx.world.creatures.spawn_init(
            CreatureInit(
                origin_template_id=0,
                pos=call.pos,
                heading=float(heading),
                phase_seed=float(phase_seed),
                type_id=call.type_id,
                flags=flags,
                ai_mode=CreatureAiMode.CHASE_PLAYER,
                health=1.0,
                max_health=1.0,
                move_speed=float(move_speed),
                reward_value=1.0,
                size=float(size),
                contact_damage=100.0,
                tint=call.tint_rgba.to_tuple(),
            ),
        )
        if creature_idx is None:
            continue
        active_mask = [bool(entry.active) for entry in ctx.world.creatures.entries]
        typo.names.assign_random(
            int(creature_idx),
            ctx.world.state.rng,
            score_xp=int(ctx.world.players[0].experience) if ctx.world.players else 0,
            active_mask=active_mask,
            dictionary_words=typo.dictionary_words,
            highscore_names=typo.highscore_names,
        )


def typo_post_step(ctx: PostStepContext) -> None:
    state = ctx.world.state
    state.bonuses.weapon_power_up = 0.0
    state.bonuses.reflex_boost = 0.0
    state.time_scale_active = False
    state.bonus_pool.reset()


def typo_input_transform(world: WorldState, inputs: list[PlayerInput]) -> list[PlayerInput]:
    if not inputs:
        world.state.typo.pending_fire_target = None
        world.state.typo.pending_reload = False
        return []

    typo = world.state.typo
    primary = inputs[0]
    aim = primary.aim
    fire_down = bool(primary.fire_down)
    fire_pressed = bool(primary.fire_pressed)
    reload_pressed = bool(primary.reload_pressed)

    if typo.pending_fire_target is not None:
        aim = typo.pending_fire_target
        fire_down = True
        fire_pressed = True
    if typo.pending_reload:
        reload_pressed = True

    typo.pending_fire_target = None
    typo.pending_reload = False

    transformed_primary = msgspec.structs.replace(
        primary,
        move=Vec2(),
        aim=Vec2(float(aim.x), float(aim.y)),
        fire_down=bool(fire_down),
        fire_pressed=bool(fire_pressed),
        reload_pressed=bool(reload_pressed),
        reload_down=False,
    )
    return [transformed_primary]
